### PR TITLE
Add new run configuration for Cosmos serverless spark

### PR DESCRIPTION
**Pull request by konjac**

### CODE REVIEW

**Comment by konjac** [comment]
@Eskibear Jenkins build failed on new dependency issue. Would you help to have a look?

**Comment by Eskibear** [comment]
AI lib 2.2.1 picked automatically. It's fine. See https://github.com/Microsoft/ApplicationInsights-Java/releases/tag/2.2.1

**Comment by Eskibear** [comment]
retest this please

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        override val module: CosmosServerlessSparkConfigurationModule,
+        cosmosServerlessSparkConfigurationFactory: CosmosServerlessSparkConfigurationFactory
+)
+    : CosmosSparkRunConfiguration(name, module, cosmosServerlessSparkConfigurationFactory) {
+    override fun getConfigurationEditor(): SettingsEditor<out RunConfiguration> {
+        return LivySparkRunConfigurationSettingsEditor(CosmosServerlessSparkConfigurable(module.project))
+    }
+
+    override fun getSuggestedNamePrefix() : String {
+        return "[Spark Job]"
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfiguration.kt]: [Cosmos Serverless Spark]?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt [hunk 2]
```diff
@@ -228,11 +236,29 @@ open class SparkSubmissionContentPanel : JPanel() {
...
 
-    val storageWithUploadPathPanel: SparkSubmissionJobUploadStorageWithUploadPathPanel =
+    var isStorageWithUploadPathPanelEnabled : Boolean = true
+
+    var storageWithUploadPathPanel: SparkSubmissionJobUploadStorageWithUploadPathPanel? =
         SparkSubmissionJobUploadStorageWithUploadPathPanel()
 
     val clusterSelectedSubject: BehaviorSubject<String> = BehaviorSubject.create<String>()
 
+    protected val sparkEventsLabel = JLabel("Spark Events:").apply {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt]: Can move these into Cosmos Serverless Spark?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java [hunk 3]
```diff
@@ -1,377 +1,388 @@
-/*
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java]: Could you help to clean up the file?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmissionParameter.java [hunk 4]
```diff
@@ -64,6 +64,7 @@
     private List<String> jars = new ArrayList<>();
     private List<String> args = new ArrayList<>();
     private Map<String, Object> jobConfig = new HashMap<>();
+    private String sparkEventsRelativePath = "spark-events";
```
Comment (t-rufang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmissionParameter.java]: Please revert change of this file. Class `SparkSubmissionParameter` is designed to prepare parameters for HDInsight spark batch job. Actually we have a class `CreateSparkBatchJobParameters`, which is designed for Cosmos serverless spark batch job.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkSubmitModel.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package com.microsoft.azure.hdinsight.spark.run.configuration
+
+import com.intellij.openapi.project.Project
+import com.microsoft.azure.hdinsight.spark.common.ServerlessSparkSubmitModel
+import com.microsoft.azure.hdinsight.spark.common.SparkSubmissionParameter
+import com.microsoft.azuretools.utils.Pair
+import java.util.stream.Stream
+
+class CosmosServerlessSparkSubmitModel(project: Project) : ServerlessSparkSubmitModel(project) {
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkSubmitModel.kt]: Maybe `ServerlessSparkSubmitModel` should be renamed to `CosmosSparkSubmitModel` in anohter PR #2347

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java [hunk 6]
```diff
@@ -287,6 +287,16 @@ public synchronized void setTableModel(@NotNull SubmissionTableModel tableModel)
         return Arrays.stream(SparkSubmissionParameter.defaultParameters);
     }
 
+    @Attribute("sparkevents_relativepath")
+    public String getSparkEventsRelativePath() {
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java]: Please remove `getSparkEventsRelativePath` and `setSparkEventsRelativePath` method in `SparkSubmitModel`. These two methods should be in sub-class `CosmosServerlessSparkSubmitModel`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkSubmissionPanelConfigurable.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * SOFTWARE.
+ */
+
+package com.microsoft.azure.hdinsight.spark.ui
+
+import com.intellij.openapi.project.Project
+import com.microsoft.azure.hdinsight.common.logger.ILogger
+
+class CosmosServerlessSparkSubmissionPanelConfigurable(project: Project)
+    : CosmosServerlessSparkSubmissionContentPanelConfigurable(project), ILogger {
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkSubmissionPanelConfigurable.kt]: Seems like `ILogger` is not needed here

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt [hunk 8]
```diff
@@ -263,9 +289,10 @@ open class SparkSubmissionContentPanel : JPanel() {
                     fill = FILL_HORIZONTAL
                 }
             }
-            row { c(clustersSelectionPrompt);             c(clustersListComboBox) }
+
+            row { c(clustersSelectionPrompt); c(clustersListComboBox) }
             row { c();                                    c(errorMessageLabels[ErrorMessage.ClusterName.ordinal]) { fill = FILL_NONE } }
-            row { c(artifactSelectLabel) }
+            row { c(artifactSelectLabel); }
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt]: `;` is not needed here

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt (hunk 2 above)
Comment (t-rufang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt]: Agree with Wei, SparkEvents related codes can be put into sub-class `CosmosServerlessSparkSubmissionContentPanel`

---
In reply to: [234097840](https://github.com/Microsoft/azure-tools-for-java/pull/2353#discussion_r234097840) [](ancestors = 234097840)

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java [hunk 9]
```diff
@@ -1,377 +1,388 @@
...
+
+            @Nullable
+            @Override
+            public IClusterDetail getClusterDetail() {
+                return getClusterDetails().stream()
+                        .filter(clusterDetail -> clusterDetail.getName().equals(getClusterName()))
+                        .findFirst()
+                        .orElse(null);
+            }
+        };
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java]: Too many blank lines here

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/SparkServerlessClusterOpsCtrl.java [hunk 10]
```diff
@@ -139,6 +142,47 @@ public SparkServerlessClusterOpsCtrl(@NotNull SparkServerlessClusterOps sparkSer
                         log().error(ex.getMessage());
                     }
                 }, ex -> log().error(ex.getMessage(), ex));
+
+        this.sparkServerlessClusterOps.getServerlessSubmitAction()
+                .observeOn(ideSchedulers.dispatchUIThread())
+                .subscribe(accountNodePair -> {
+                    log().info(String.format("Submit message received. cluster: %s, node: %s",
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/SparkServerlessClusterOpsCtrl.java]: "Submit message received. account: %s, node: %s"

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/SparkServerlessClusterOpsCtrl.java [hunk 11]
```diff
@@ -139,6 +142,47 @@ public SparkServerlessClusterOpsCtrl(@NotNull SparkServerlessClusterOps sparkSer
...
+                            accountNodePair.getLeft(), accountNodePair.getRight()));
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/SparkServerlessClusterOpsCtrl.java]:
```
                    log().info(String.format("Submit message received. account: %s, node: %s",
                            accountNodePair.getLeft().getName(), accountNodePair.getRight().getName()));
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/SparkServerlessClusterOpsCtrl.java [hunk 12]
```diff
@@ -139,6 +142,47 @@ public SparkServerlessClusterOpsCtrl(@NotNull SparkServerlessClusterOps sparkSer
...
+                                        new CosmosServerlessSparkConfigurationFactory(new CosmosServerlessSparkConfigurationType())));
+
+                        context.putData(RUN_CONFIGURATION_SETTING, runConfigurationSetting)
+                                .putData(CLUSTER, adlAccount);
+
+                        Presentation actionPresentation = new Presentation("Submit Cosmos Serverless Spark Job");
+                        actionPresentation.setDescription("Submit specified Spark application into the remote cluster");
+
+                        AnActionEvent event = AnActionEvent.createFromDataContext(
+                                String.format("Azure Data Lake Spark pool %s:%s context menu",
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/SparkServerlessClusterOpsCtrl.java]: Maybe "Azure Data Lake Spark pool" should be "Cosmos Serverless Cluster"?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/rest/azure/serverless/spark/models/AnalyticsActivity.java [hunk 13]
```diff
@@ -27,10 +27,12 @@
 import java.util.UUID;
 import java.util.Map;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 
 /**
  * The common Data Lake Analytics activity information.
  */
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (t-rufang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/rest/azure/serverless/spark/models/AnalyticsActivity.java]: Maybe we should rebase the repo first since json ignore is added in the latest develop branch

File: Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/sparkserverlessnode/SparkServerlessADLAccountNode.java [hunk 14]
```diff
@@ -80,6 +80,9 @@ protected void refreshItems() throws AzureCmdException {
     protected void loadActions() {
         super.loadActions();
 
+        addAction("Submit Serverless Spark Job", new CosmosSparkServerlessSubmitAction(
+                this, adlAccount, SparkServerlessClusterOps.getInstance().getServerlessSubmitAction()));
```
Comment (t-rufang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/sparkserverlessnode/SparkServerlessADLAccountNode.java]: I think we should sync with Wei to make sure whether we need a switch for this feature

File: PluginsAndFeatures/azure-toolkit-for-intellij/resources/META-INF/plugin.xml [hunk 15]
```diff
@@ -316,6 +316,7 @@
     <configurationType implementation="com.microsoft.intellij.runner.webapp.WebAppConfigurationType"/>
     <configurationType implementation="com.microsoft.azure.hdinsight.spark.run.configuration.LivySparkBatchJobRunConfigurationType"/>
     <configurationType implementation="com.microsoft.azure.hdinsight.spark.run.configuration.CosmosSparkConfigurationType"/>
+    <configurationType implementation="com.microsoft.azure.hdinsight.spark.run.configuration.CosmosServerlessSparkConfigurationType"/>
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/resources/META-INF/plugin.xml]: We can comment this line for this release.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfiguration.kt [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package com.microsoft.azure.hdinsight.spark.run.configuration
+
+import com.intellij.execution.configurations.RunConfiguration
+import com.intellij.openapi.options.SettingsEditor
+
+class CosmosServerlessSparkConfiguration(
+        name: String,
+        override val module: CosmosServerlessSparkConfigurationModule,
+        cosmosServerlessSparkConfigurationFactory: CosmosServerlessSparkConfigurationFactory
+)
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfiguration.kt]: Could you put it to the last line ending?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt [hunk 17]
```diff
@@ -263,31 +272,44 @@ open class SparkSubmissionContentPanel : JPanel() {
...
-            row {   c();                                  c(errorMessageLabels[ErrorMessage.LocalArtifact.ordinal]) { fill = FILL_NONE }}
-            row { c(mainClassPrompt);                     c(mainClassTextField) }
-            row { c();                                    c(errorMessageLabels[ErrorMessage.MainClass.ordinal]) }
-            row { c(jobConfigPrompt);                     c(jobConfTablScrollPane) }
-            row { c();                                    c(errorMessageLabels[ErrorMessage.JobConfiguration.ordinal]) }
-            row { c(commandLineArgsPrompt);               c(commandLineTextField) }
-            row { c(refJarsPrompt);                       c(referencedJarsTextField) }
-            row { c(refFilesPrompt);                      c(referencedFilesTextField) }
-            row { c(storageWithUploadPathPanel) { colSpan = 2; fill = FILL_HORIZONTAL }; }
+            row { c(ideaArtifactPrompt) { indent = 1 }; c(selectedArtifactComboBox) }
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt]: Could you restore it?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/CosmosServerlessSparkSubmissionContentPanelConfigurable.kt [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+import com.intellij.openapi.project.Project
+import com.microsoft.azure.hdinsight.sdk.cluster.IClusterDetail
+import com.microsoft.azure.hdinsight.sdk.common.azure.serverless.AzureSparkServerlessClusterManager
+import com.microsoft.azure.hdinsight.spark.common.SparkSubmitModel
+import com.microsoft.azure.hdinsight.spark.run.configuration.CosmosServerlessSparkSubmitModel
+import com.microsoft.azuretools.azurecommons.helpers.NotNull
+import rx.Observable
+import javax.swing.JComponent
+
+open class CosmosServerlessSparkSubmissionContentPanelConfigurable : SparkSubmissionContentPanelConfigurable {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/CosmosServerlessSparkSubmissionContentPanelConfigurable.kt]: Let's combine this with `CosmosServerlessSparkSubmissionContentPanel`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt [hunk 19]
```diff
@@ -277,10 +284,20 @@ open class SparkSubmissionContentPanel : JPanel() {
             row { c(commandLineArgsPrompt);               c(commandLineTextField) }
             row { c(refJarsPrompt);                       c(referencedJarsTextField) }
             row { c(refFilesPrompt);                      c(referencedFilesTextField) }
-            row { c(storageWithUploadPathPanel) { colSpan = 2; fill = FILL_HORIZONTAL }; }
+
+            if (enableStoragePanel) {
+                row { c(storageWithUploadPathPanel) { colSpan = 2; fill = FILL_HORIZONTAL }; }
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt]: Can the `storageWithUploadPathPanel` be overridden as an empty `JPanel()`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt [hunk 20]
```diff
@@ -277,10 +284,20 @@ open class SparkSubmissionContentPanel : JPanel() {
...
+            }
         }
+        return formBuilder.buildPanel()
+    }
 
-        this.add(formBuilder.buildPanel())
+    private val sparkSubmissionPanel: JPanel by lazy {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanel.kt]: Is it a must-be change?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java [hunk 21]
```diff
@@ -63,18 +63,21 @@
     @NotNull
     private final Project myProject;
 
-    private SparkSubmissionContentPanel submissionPanel = new SparkSubmissionContentPanel();
+    protected SparkSubmissionContentPanel submissionPanel;
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java]: Could you add an annotation `@Nullable`?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/sparkserverlessnode/SparkServerlessADLAccountNode.java [hunk 22]
```diff
@@ -80,6 +80,9 @@ protected void refreshItems() throws AzureCmdException {
     protected void loadActions() {
         super.loadActions();
 
+        /*addAction("Submit Serverless Spark Job", new CosmosSparkServerlessSubmitAction(
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/sparkserverlessnode/SparkServerlessADLAccountNode.java]: Add TODO comment

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java [hunk 23]
```diff
@@ -63,18 +63,21 @@
...
     private SparkSubmissionJobUploadStorageCtrl jobUploadStorageCtrl;
 
     // Cluster refresh publish subject with preselected cluster name as event
     @Nullable
     private BehaviorSubject<String> clustersRefreshSub;
 
     public SparkSubmissionContentPanelConfigurable(@NotNull Project project) {
-        this.myProject = project;
+        this(project, true);
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java]: Could you also revert these changes?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java (hunk 21 above)
Comment (konjac) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java]: this line has been reverted.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/sparkserverlessnode/SparkServerlessADLAccountNode.java (hunk 22 above)
Comment (konjac) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/sparkserverlessnode/SparkServerlessADLAccountNode.java]: Added.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java [hunk 24]
```diff
@@ -71,6 +71,10 @@
     private BehaviorSubject<String> clustersRefreshSub;
 
     public SparkSubmissionContentPanelConfigurable(@NotNull Project project) {
+        this(project, true);
+    }
+
+    public SparkSubmissionContentPanelConfigurable(@NotNull Project project, boolean enableStoragePanel) {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java]: Still be here...
